Sources/Noora/Components/CLIProgressBar.swift

```diff
@@ -0,0 +1,116 @@
+import Foundation
+import Rainbow
+
+
+struct CLIProgressBar {
+    
+    let message: String
+    let successMessage: String?
+    let errorMessage: String?
+    let total: Int
+    let action: (@escaping (String) -> Void) async throws -> Void
+    let theme: Theme
+    let terminal: Terminaling
+    let renderer: Rendering
+    let standardPipelines: StandardPipelines
+    var progressBar: ProgressBar
+    
+    init(
+        message: String,
+        successMessage: String?,
+        errorMessage: String?,
+        total: Int,
+        action: @escaping (@escaping (String) -> Void) async throws -> Void,
+        theme: Theme,
+        terminal: Terminaling,
+        renderer: Rendering,
+        standardPipelines: StandardPipelines,
+        progressBar: ProgressBar = DefaultProgressBar()
+    ) {
+        self.message = message
+        self.successMessage = successMessage
+        self.errorMessage = errorMessage
+        self.total = total
+        self.action = action
+        self.theme = theme
+        self.terminal = terminal
+        self.renderer = renderer
+        self.standardPipelines = standardPipelines
+        self.progressBar = progressBar
+    }
+    
+    func run() async throws {
+        if terminal.isInteractive {
+            try await runInteractive()
+        } else {
+            try await runNonInteractive()
+        }
+    }
+    
+    func runInteractive() async throws {
+
+        var bar: String = ""
+        var progressPercentage = 0
+        var lastMessage = formatMessage(message)
+
+        progressBar.startProgress(total: total, interval: 0.05) { progressBarState, percentage in
+            bar = progressBarState
+            progressPercentage = percentage
+            self.render(lastMessage, bar, progressPercentage)
+        }
```

Comment:
As mentioned in another comment, it shouldn't be the controller the one that controls the progress, but the caller that uses the component to run an asynchronous action.